// @hyperledger/fabric-gateway/dist/commit.js
"use strict";
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommitImpl = void 0;
const status_1 = require("./status");
class CommitImpl {
    #client;
    #signingIdentity;
    #transactionId;
    #signedRequest;
    constructor(options) {
        this.#client = options.client;
        this.#signingIdentity = options.signingIdentity;
        this.#transactionId = options.transactionId;
        this.#signedRequest = options.signedRequest;
    }
    getBytes() {
        return this.#signedRequest.serializeBinary();
    }
    getDigest() {
        const request = this.#signedRequest.getRequest_asU8();
        return this.#signingIdentity.hash(request);
    }
    async getStatus(options) {
        await this.#sign();
        const response = await this.#client.commitStatus(this.#signedRequest, options);
        return this.#newStatus(response);
    }
    getTransactionId() {
        return this.#transactionId;
    }
    setSignature(signature) {
        this.#signedRequest.setSignature(signature);
    }
    async #sign() {
        if (this.#isSigned()) {
            return;
        }
        const signature = await this.#signingIdentity.sign(this.getDigest());
        this.setSignature(signature);
    }
    #isSigned() {
        const signatureLength = this.#signedRequest.getSignature()?.length || 0;
        return signatureLength > 0;
    }
    #newStatus(response) {
        const code = response.getResult() ?? status_1.StatusCode.INVALID_OTHER_REASON;
        return {
            blockNumber: BigInt(response.getBlockNumber()),
            code,
            successful: code === status_1.StatusCode.VALID,
            transactionId: this.#transactionId,
        };
    }
}
exports.CommitImpl = CommitImpl;
