// @hyperledger/fabric-gateway/dist/inmemorycheckpointer.js
"use strict";
/*
 * Copyright 2022 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryCheckPointer = void 0;
/**
 * In-memory checkpointer class used to persist checkpointer state in memory.
 */
class InMemoryCheckPointer {
    #blockNumber;
    #transactionId;
    checkpointBlock(blockNumber) {
        this.#blockNumber = blockNumber + BigInt(1);
        this.#transactionId = undefined;
        return Promise.resolve();
    }
    checkpointTransaction(blockNumber, transactionId) {
        this.#blockNumber = blockNumber;
        this.#transactionId = transactionId;
        return Promise.resolve();
    }
    async checkpointChaincodeEvent(event) {
        await this.checkpointTransaction(event.blockNumber, event.transactionId);
    }
    getBlockNumber() {
        return this.#blockNumber;
    }
    getTransactionId() {
        return this.#transactionId;
    }
}
exports.InMemoryCheckPointer = InMemoryCheckPointer;
